gravitee-am-factor/gravitee-am-factor-api/src/main/java/io/gravitee/am/factor/api/OTPFactorProvider.java

```diff
@@ -62,11 +62,13 @@ protected Completable verifyOTP(EnrolledFactor enrolledFactor, int returnDigits,
             try {
                 final String otpCode = generateOTP(enrolledFactor, returnDigits);
                 if (!code.equals(otpCode)) {
+                    log.debug("Invalid 2FA code, not the same value");
                     emitter.onError(invalid2faCodeException());
                 }
                 // get last connection date of the user to test code
                 Long expireAt = enrolledFactor.getSecurity().getData(FactorDataKeys.KEY_EXPIRE_AT, Long.class);
                 if (expireAt == null || Instant.now().isAfter(Instant.ofEpochMilli(expireAt))) {
+                    log.debug("Invalid 2FA code, invalid expire at");
```

Comment:
Might be good to add the value of the expiry date in the log : 

```
 Long expireAt = enrolledFactor.getSecurity().getData(FactorDataKeys.KEY_EXPIRE_AT, Long.class);
 Instant now = Instant.now();
 if (expireAt == null || now.isAfter(Instant.ofEpochMilli(expireAt))) {
  long differenceMillis = now.toEpochMilli() - expInstant.toEpochMilli();
  String msg = "Invalid 2FA code, expiry date " + expireAt + ". Current time: " + now + ", a difference of " +
                        differenceMillis + " milliseconds.";
  log.debug(msg);
  emitter.onError(new InvalidCodeException("Invalid 2FA Code"));
}
```

Reply:
done